### PR TITLE
Property test arbitrary depth variant

### PR DESCRIPTION
**Pull request by bdemann**

### CODE REVIEW

File: property_tests/arbitraries/candid/constructed/record_arb/base.ts [hunk 1]
```diff
@@ -0,0 +1,110 @@
...
+import { JsFunctionNameArb } from '../../../js_function_name_arb';
+import { Record } from './index';
+
+type Field = [string, CandidMeta<CandidType>];
+
+export function RecordArb(candidTypeArb: fc.Arbitrary<CandidMeta<CandidType>>) {
+    return fc
+        .tuple(
+            UniqueIdentifierArb('typeDeclaration'),
+            fc.uniqueArray(fc.tuple(JsFunctionNameArb, candidTypeArb), {
```
Comment (lastmjs) [inline property_tests/arbitraries/candid/constructed/record_arb/base.ts:15]: Is JsFunctionNameArb unique?

Comment (lastmjs) [reply property_tests/arbitraries/candid/constructed/record_arb/base.ts:15]: Is that what uniqueArray is doing somehow? What happens if there are two JsFunctionNameArbs with the same name? Seems like we should be using UniqueIdentifierArb

File: property_tests/arbitraries/candid/constructed/variant_arb/base.ts [hunk 2]
```diff
@@ -0,0 +1,125 @@
...
+
+function VariantFieldsArb(
+    candidTypeArb: fc.Arbitrary<CandidMeta<CandidType>>
+): fc.Arbitrary<Field[]> {
+    return fc.uniqueArray(fc.tuple(JsFunctionNameArb, candidTypeArb), {
+        selector: (entry) => entry[0],
+        minLength: 1
+        // Although no minLength is technically required (according to the
+        // spec), the DFX CLI itself currently errors out trying to pass
+        // an empty object.
```
Comment (lastmjs) [inline property_tests/arbitraries/candid/constructed/variant_arb/base.ts:18]: Is this still a problem? We should open an issue somewhere

File: property_tests/arbitraries/candid/candid_arb.ts [hunk 3]
```diff
@@ -1,16 +1,18 @@
 import fc from 'fast-check';
-import { deepEqual } from 'fast-equals';
 import { CandidType } from './candid_type_arb';
 
 // TODO we're thinking that Candid is not the best name for this. What is better?
-export type CandidMeta<T extends CandidType> = {
+export type CandidMeta<T extends CandidType, E = T> = {
     value: T;
```
Comment (lastmjs) [inline property_tests/arbitraries/candid/candid_arb.ts:6]: Can we think of a better name for this, somehow to distinguish these values? It's just not clear what the value is and why it's different than the expected value...do you think this is important?

For example should we say agentArgumentValue and agentResponseValue or something?

File: property_tests/tests/tuple/test/test.ts [hunk 4]
```diff
@@ -94,12 +95,19 @@ function generateBody(
     `;
 }
 
+function replacer(key: any, value: any) {
```
Comment (lastmjs) [inline property_tests/tests/tuple/test/test.ts:98]: What is this replacer for? is there a JSON.stringify somewhere? We should use the replacer and reviver I've created now, perhaps we should open an issue for this or comment it in here

**Review by lastmjs — CHANGES_REQUESTED** [review]

File: property_tests/arbitraries/candid/candid_arb.ts (hunk 3 above)
Comment (lastmjs) [reply property_tests/arbitraries/candid/candid_arb.ts:6]: Change these to argumentValue and expectedResponseValue

File: property_tests/tests/tuple/test/test.ts (hunk 4 above)
Comment (lastmjs) [reply property_tests/tests/tuple/test/test.ts:98]: Remove this

File: property_tests/arbitraries/candid/constructed/variant_arb/base.ts (hunk 2 above)
Comment (bdemann) [reply property_tests/arbitraries/candid/constructed/variant_arb/base.ts:18]: We should also open an issue for the tuple inconsistency we found

File: property_tests/arbitraries/candid/constructed/record_arb/base.ts (hunk 1 above)
Comment (bdemann) [reply property_tests/arbitraries/candid/constructed/record_arb/base.ts:15]: Unique Identifier Arb is going to make sure that globally all of the names are unique, for this case we don't need to them to be globally unique, we just need them to be locally unique to this particular record, another record could have fields of the same name and that would be okay. uniqueArray is what we want here.

**Review by lastmjs — APPROVED** [review]

**Pull request merged**